File: plugins/org.eclipse.tm.tcf.core/src/org/eclipse/tm/internal/tcf/services/remote/MemoryProxy.java
/*******************************************************************************
 * Copyright (c) 2007, 2008 Wind River Systems, Inc. and others.
 * All rights reserved. This program and the accompanying materials 
 * are made available under the terms of the Eclipse Public License v1.0 
 * which accompanies this distribution, and is available at 
 * http://www.eclipse.org/legal/epl-v10.html 
 *  
 * Contributors:
 *     Wind River Systems - initial API and implementation
 *******************************************************************************/
package org.eclipse.tm.internal.tcf.services.remote;

import java.io.IOException;
import java.math.BigInteger;
import java.util.Arrays;
import java.util.Collection;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

import org.eclipse.tm.tcf.core.Command;
import org.eclipse.tm.tcf.protocol.IChannel;
import org.eclipse.tm.tcf.protocol.IErrorReport;
import org.eclipse.tm.tcf.protocol.IToken;
import org.eclipse.tm.tcf.protocol.JSON;
import org.eclipse.tm.tcf.services.IMemory;


public class MemoryProxy implements IMemory {

    private final IChannel channel;
    private final Map<MemoryListener,IChannel.IEventListener> listeners =
        new HashMap<MemoryListener,IChannel.IEventListener>();
    
    private static class Range implements Comparable<Range> {
        int offs;
        int size;
        int stat;
        String msg;
        
        public int compareTo(Range o) {
            if (offs < o.offs) return -1;
            if (offs > o.offs) return +1;
            return 0;
        }
    }
    
    private class MemoryErrorReport extends MemoryError implements ErrorOffset, IErrorReport {
        
        private static final long serialVersionUID = 796525409870265390L;
        private final Map<String,Object> attrs;
        private final Range[] ranges;
        
        @SuppressWarnings("unchecked")
        MemoryErrorReport(String msg, Map<String,Object> attrs, Number addr, Object ranges) {
            super(msg);
            this.attrs = attrs;
            Collection<Map<String,Object>> c = (Collection<Map<String,Object>>)ranges;
            this.ranges = c == null ? null : new Range[c.size()];
            if (c != null) {
                int n = 0;
                BigInteger addr_bi = addr instanceof BigInteger ?
                        (BigInteger)addr : new BigInteger(addr.toString());
                for (Map<String,Object> m : c) {
                    Range r = new Range();
                    Number x = (Number)m.get("addr");
                    BigInteger y = x instanceof BigInteger ?
                            (BigInteger)x : new BigInteger(x.toString());
                    r.offs = addr_bi.subtract(y).intValue();
                    r.size = ((Number)m.get("size")).intValue();
                    r.stat = ((Number)m.get("stat")).intValue();
                    r.msg = Command.toErrorString(m.get("msg"));
                    assert r.offs >= 0;
                    assert r.size >= 0;
                    this.ranges[n++] = r;
                }
                Arrays.sort(this.ranges);
            }
        }

        public int getErrorCode() {
            Number n = (Number)attrs.get(ERROR_CODE);
            if (n == null) return 0;
            return n.intValue();
        }

        public int getAltCode() {
            Number n = (Number)attrs.get(ERROR_ALT_CODE);
            if (n == null) return 0;
            return n.intValue();
        }

        public String getAltOrg() {
            return (String)attrs.get(ERROR_ALT_ORG);
        }

        public Map<String, Object> getAttributes() {
            return attrs;
        }

        public String getMessage(int offset) {
            if (ranges == null) return null;
            int l = 0;
            int h = ranges.length - 1;
            while (l <= h) {
                int n = (l + h) / 2;
                Range r = ranges[n];
                if (r.offs > offset) {
                    h = n - 1;
                }
                else if (offset >= r.offs + r.size) {
                    l = n + 1;
                }
                else {
                    return r.msg;
                }
            }
            return null;
        }

        public int getStatus(int offset) {
            if (ranges == null) return BYTE_UNKNOWN;
            int l = 0;
            int h = ranges.length - 1;
            while (l <= h) {
                int n = (l + h) / 2;
                Range r = ranges[n];
                if (r.offs > offset) {
                    h = n - 1;
                }
                else if (offset >= r.offs + r.size) {
                    l = n + 1;
                }
                else {
                    return r.stat;
                }
            }
            return BYTE_UNKNOWN;
        }
    }

    private class MemContext implements MemoryContext {

        private final Map<String,Object> props;

        MemContext(Map<String,Object> props) {
            this.props = props;
        }

        public String getID() {
            return (String)props.get(PROP_ID);
        }

        public String getParentID() {
            return (String)props.get(PROP_PARENT_ID);
        }

        public int getAddressSize() {
            Number n = (Number)props.get(PROP_ADDRESS_SIZE);
            if (n == null) return 0;
            return n.intValue();
        }

        public String getProcessID() {
            return (String)props.get(PROP_PROCESS_ID);
        }

        public boolean isBigEndian() {
            Boolean n = (Boolean)props.get(PROP_BIG_ENDIAN);
            if (n == null) return false;
            return n.booleanValue();
        }

        @SuppressWarnings("unchecked")
        public Collection<String> getAccessTypes() {
            return (Collection<String>)props.get(PROP_ACCESS_TYPES);
        }

        public Number getEndBound() {
            return (Number)props.get(PROP_END_BOUND);
        }

        public String getName() {
            return (String)props.get(PROP_NAME);
        }

        public Number getStartBound() {
            return (Number)props.get(PROP_START_BOUND);
        }

        public Map<String, Object> getProperties() {
            return props;
        }

        public IToken fill(final Number addr, int word_size,
                byte[] value, int size, int mode, final DoneMemory done) {
            return new MemoryCommand("fill", new Object[] {
                getID(), addr, word_size, size, mode, value
            } ) {
                public void done(Exception error, Object[] args) {
                    MemoryError e = null;
                    if (error != null) {
                        e = new MemoryError(error.getMessage());
                    }
                    else {
                        assert args.length == 2;
                        e = toMemoryError(addr, args[0], args[1]);
                    }
                    done.doneMemory(token, e);
                }
            }.token;
        }

        public IToken get(final Number addr, int word_size,
                final byte[] buf, final int offs, final int size,
                int mode, final DoneMemory done) {
            return new MemoryCommand("get", new Object[] {
                    getID(), addr, word_size, size, mode
                } ) {
                    public void done(Exception error, Object[] args) {
                        MemoryError e = null;
                        if (error != null) {
                            e = new MemoryError(error.getMessage());
                        }
                        else {
                            assert args.length == 3;
                            JSON.toByteArray(buf, offs, size, args[0]);
                            e = toMemoryError(addr, args[1], args[2]);
                        }
                        done.doneMemory(token, e);
                    }
                }.token;
        }

        public IToken set(final Number addr, int word_size,
                byte[] buf, int offs, int size, int mode, final DoneMemory done) {
            return new MemoryCommand("set", new Object[] {
                    getID(), addr, word_size, size, mode, new JSON.Binary(buf, offs, size)
                } ) {
                    public void done(Exception error, Object[] args) {
                        MemoryError e = null;
                        if (error != null) {
                            e = new MemoryError(error.getMessage());
                        }
                        else {
                            assert args.length == 2;
                            e = toMemoryError(addr, args[0], args[1]);
                        }
                        done.doneMemory(token, e);
                    }
                }.token;
        }
        
        public String toString() {
            return "[Memory Context " + props.toString() + "]";
        }
    }

    public MemoryProxy(IChannel channel) {
        this.channel = channel;
    }

    public void addListener(final MemoryListener listener) {
        IChannel.IEventListener l = new IChannel.IEventListener() {

            public void event(String name, byte[] data) {
                try {
                    Object[] args = JSON.parseSequence(data);
                    if (name.equals("contextAdded")) {
                        assert args.length == 1;
                        listener.contextAdded(toContextArray(args[0]));
                    }
                    else if (name.equals("contextChanged")) {
                        assert args.length == 1;
                        listener.contextChanged(toContextArray(args[0]));
                    }
                    else if (name.equals("contextRemoved")) {
                        assert args.length == 1;
                        listener.contextRemoved(toStringArray(args[0]));
                    }
                    else if (name.equals("memoryChanged")) {
                        assert args.length == 2;
                        listener.memoryChanged((String)args[0],
                                toAddrArray(args[1]), toSizeArray(args[1]));
                    }
                    else {
                        throw new IOException("Memory service: unknown event: " + name);
                    }
                }
                catch (Throwable x) {
                    channel.terminate(x);
                }
            }
        };
        channel.addEventListener(this, l);
        listeners.put(listener, l);
    }

    public void removeListener(MemoryListener listener) {
        IChannel.IEventListener l = listeners.remove(listener);
        if (l != null) channel.removeEventListener(this, l);
    }

    public IToken getContext(String context_id, final DoneGetContext done) {
        return new Command(channel, this, "getContext", new Object[]{ context_id }) {
            @SuppressWarnings("unchecked")
            @Override
            public void done(Exception error, Object[] args) {
                MemContext ctx = null;
                if (error == null) {
                    assert args.length == 2;
                    error = toError(args[0]);
                    if (args[1] != null) ctx = new MemContext((Map<String,Object>)args[1]);
                }
                done.doneGetContext(token, error, ctx);
            }
        }.token;
    }
    
    public IToken getChildren(String parent_context_id, final DoneGetChildren done) {
        return new Command(channel, this, "getChildren", new Object[]{ parent_context_id }) {
            @Override
            public void done(Exception error, Object[] args) {
                String[] arr = null;
                if (error == null) {
                    assert args.length == 2;
                    error = toError(args[0]);
                    arr = toStringArray(args[1]);
                }
                done.doneGetChildren(token, error, arr);
            }
        }.token;
    }

    public String getName() {
        return NAME;
    }
    
    private abstract class MemoryCommand extends Command {
        
        MemoryCommand(String cmd, Object[] args) {
            super(channel, MemoryProxy.this, cmd, args);
        }

        @SuppressWarnings("unchecked")
        MemoryError toMemoryError(Number addr, Object data, Object ranges) {
            if (data == null) return null;
            Map<String,Object> map = (Map<String,Object>)data;
            Integer code = (Integer)map.get(IErrorReport.ERROR_CODE);
            String cmd = getCommandString();
            if (cmd.length() > 72) cmd = cmd.substring(0, 72) + "...";
            MemoryError e = new MemoryErrorReport(
                    "TCF command exception:" +
                    "\nCommand: " + cmd +
                    "\nException: " + toErrorString(data) +
                    "\nError code: " + code,
                    map, addr, ranges);
            Object caused_by = map.get(IErrorReport.ERROR_CAUSED_BY);
            if (caused_by != null) e.initCause(toError(caused_by, false));
            return e;
        }
    }

    @SuppressWarnings("unchecked")
    private MemoryContext[] toContextArray(Object o) {
        Collection<Map<String,Object>> c = (Collection<Map<String,Object>>)o;
        if (c == null) return new MemoryContext[0];
        int n = 0;
        MemoryContext[] ctx = new MemoryContext[c.size()];
        for (Iterator<Map<String,Object>> i = c.iterator(); i.hasNext();) {
            ctx[n++] = new MemContext(i.next());
        }
        return ctx;
    }

    @SuppressWarnings("unchecked")
    private long[] toSizeArray(Object o) {
        Collection<Map<String,Object>> c = (Collection<Map<String,Object>>)o;
        if (c == null) return null;
        long[] a = new long[c.size()];
        int n = 0;
        for (Map<String,Object> m : c) {
            Number sz = (Number)m.get("size");
            a[n++] = sz == null ? 0 : sz.longValue();
        }
        return a;
    }

    @SuppressWarnings("unchecked")
    private Number[] toAddrArray(Object o) {
        Collection<Map<String,Object>> c = (Collection<Map<String,Object>>)o;
        if (c == null) return null;
        Number[] a = new Number[c.size()];
        int n = 0;
        for (Map<String,Object> m : c) {
            a[n++] = (Number)m.get("addr");
        }
        return a;
    }

    @SuppressWarnings("unchecked")
    private String[] toStringArray(Object o) {
        if (o == null) return null;
        Collection<String> c = (Collection<String>)o;
        return (String[])c.toArray(new String[c.size()]);
    }
}

File: python/src/tcf/services/remote/MemoryProxy.py
# *******************************************************************************
# * Copyright (c) 2011 Wind River Systems, Inc. and others.
# * All rights reserved. This program and the accompanying materials
# * are made available under the terms of the Eclipse Public License v1.0
# * which accompanies this distribution, and is available at
# * http://www.eclipse.org/legal/epl-v10.html
# *
# * Contributors:
# *     Wind River Systems - initial API and implementation
# *******************************************************************************

from tcf import errors, channel
from tcf.services import memory
from tcf.channel.Command import Command

class Range(object):
    offs = 0
    size = 0
    stat = 0
    msg = None
    def __cmp__(self, o):
        if self.offs < o.offs: return -1
        if self.offs > o.offs: return +1
        return 0

class MemoryErrorReport(errors.ErrorReport, memory.MemoryError, memory.ErrorOffset):
    def __init__(self, msg, attrs, addr, ranges):
        super(MemoryErrorReport, self).__init__(msg, attrs)
        if ranges is None:
            self.ranges = None
        else:
            self.ranges = []
            for m in ranges:
                r = Range()
                x = m.get(memory.ErrorOffset.RANGE_KEY_ADDR)
                if isinstance(x, str):
                    y = int(x)
                else:
                    y = x
                r.offs = y - addr
                r.size = m.get(memory.ErrorOffset.RANGE_KEY_SIZE)
                r.stat = m.get(memory.ErrorOffset.RANGE_KEY_STAT)
                r.msg = errors.toErrorString(m.get(memory.ErrorOffset.RANGE_KEY_MSG))
                assert r.offs >= 0
                assert r.size >= 0
                self.ranges.append(r)
            self.ranges.sort()

    def getMessage(self, offset):
        if self.ranges is None: return None
        l = 0
        h = len(self.ranges) - 1
        while l <= h:
            n = (l + h) / 2
            r = self.ranges[n]
            if r.offs > offset:
                h = n - 1
            elif offset >= r.offs + r.size:
                l = n + 1
            else:
                return r.msg
        return None

    def getStatus(self, offset):
        if self.ranges is None: return memory.ErrorOffset.BYTE_UNKNOWN
        l = 0
        h = len(self.ranges) - 1
        while l <= h:
            n = (l + h) / 2
            r = self.ranges[n]
            if r.offs > offset:
                h = n - 1
            elif offset >= r.offs + r.size:
                l = n + 1
            else:
                return r.stat
        return memory.ErrorOffset.BYTE_UNKNOWN


class MemContext(memory.MemoryContext):
    def __init__(self, service, props):
        super(MemContext, self).__init__(props)
        self.service = service

    def fill(self, addr, word_size, value, size, mode, done):
        service = self.service
        id = self.getID()
        done = service._makeCallback(done)
        class FillCommand(MemoryCommand):
            def __init__(self):
                super(FillCommand, self).__init__(service,
                        "fill", (id, addr, word_size, size, mode, value))
            def done(self, error, args):
                e = None
                if error:
                    e = memory.MemoryError(error.message)
                else:
                    assert len(args) == 2
                    e = self.toMemoryError(args[0], args[1])
                done.doneMemory(self.token, e)
        return FillCommand().token

    def get(self, addr, word_size, buf, offs, size, mode, done):
        service = self.service
        id = self.getID()
        done = service._makeCallback(done)
        class GetCommand(MemoryCommand):
            def __init__(self):
                super(GetCommand, self).__init__(service,
                        "get", (id, addr, word_size, size, mode))
            def done(self, error, args):
                e = None
                if error:
                    e = memory.MemoryError(error.message)
                else:
                    assert len(args) == 3
                    bytes = channel.toByteArray(args[0])
                    assert len(bytes) <= size
                    buf[offs:offs+len(bytes)] = bytes
                    e = self.toMemoryError(args[1], args[2])
                done.doneMemory(self.token, e)
        return GetCommand().token

    def set(self, addr, word_size, buf, offs, size, mode, done):
        service = self.service
        id = self.getID()
        done = service._makeCallback(done)
        class SetCommand(MemoryCommand):
            def __init__(self):
                super(SetCommand, self).__init__(service,
                        "set", (id, addr, word_size, size, mode, bytearray(buf[offs:offs:size])))
            def done(self, error, args):
                e = None
                if error:
                    e = memory.MemoryError(error.message)
                else:
                    assert len(args) == 2
                    e = self.toMemoryError(args[1], args[2])
                done.doneMemory(self.token, e)
        return SetCommand().token

class MemoryProxy(memory.MemoryService):
    def __init__(self, channel):
        self.channel = channel
        self.listeners = {}

    def getContext(self, context_id, done):
        done = self._makeCallback(done)
        service = self
        class GetContextCommand(Command):
            def __init__(self):
                super(GetContextCommand, self).__init__(service.channel, service, "getContext", (context_id,))
            def done(self, error, args):
                ctx = None
                if not error:
                    assert len(args) == 2
                    error = self.toError(args[0])
                    if args[1]: ctx = MemContext(service, args[1])
                done.doneGetContext(self.token, error, ctx)
        return GetContextCommand().token

    def getChildren(self, parent_context_id, done):
        done = self._makeCallback(done)
        service = self
        class GetChildrenCommand(Command):
            def __init__(self):
                super(GetChildrenCommand, self).__init__(service.channel, service, "getChildren", (parent_context_id,))
            def done(self, error, args):
                contexts = None
                if not error:
                    assert len(args) == 2
                    error = self.toError(args[0])
                    contexts = args[1]
                done.doneGetChildren(self.token, error, contexts)
        return GetChildrenCommand().token

    def addListener(self, listener):
        l = ChannelEventListener(self, listener)
        self.channel.addEventListener(self, l)
        self.listeners[listener] = l

    def removeListener(self, listener):
        l = self.listeners.get(listener)
        if l:
            del self.listeners[listener]
            self.channel.removeEventListener(self, l)

class MemoryCommand(Command):
    def __init__(self, service, cmd, args):
        super(MemoryCommand, self).__init__(service.channel, service, cmd, args)
    def toMemoryError(self, addr, data, ranges):
        if data is None: return None
        code = data.get(errors.ERROR_CODE)
        cmd = self.getCommandString()
        if len(cmd) > 72: cmd = cmd[0:72] + "..."
        e = MemoryErrorReport(
                "TCF command exception:\nCommand: %s\nException: %s\nError code: " % (
                    cmd, self.toErrorString(data), code),
                map, addr, ranges)
        caused_by = data.get(errors.ERROR_CAUSED_BY)
        if caused_by is not None: e.caused_by = self.toError(caused_by, False)
        return e


class ChannelEventListener(channel.EventListener):
    def __init__(self, service, listener):
        self.service = service
        self.listener = listener
    def event(self, name, data):
        try:
            args = channel.fromJSONSequence(data)
            if name == "contextAdded":
                assert len(args) == 1
                self.listener.contextAdded(_toContextArray(args[0]))
            elif name == "contextChanged":
                assert len(args) == 1
                self.listener.contextChanged(_toContextArray(args[0]))
            elif name == "contextRemoved":
                assert len(args) == 1
                self.listener.contextRemoved(args[0])
            elif name == "memoryChanged":
                assert len(args) == 2
                self.listener.memoryChanged(args[0], _toAddrArray(args[1]), _toSizeArray(args[1]))
            else:
                raise IOError("Memory service: unknown event: " + name);
        except Exception as x:
            self.service.channel.terminate(x)


def _toContextArray(o):
    if o is None: return None
    ctx = []
    for m in o: ctx.append(MemContext(m))
    return ctx

def _toSizeArray(o):
    if o is None: return None
    a = []
    for m in o:
        sz = m.get("size", 0)
        a.append(sz)
    return a

def _toAddrArray(o):
    if o is None: return None
    a = []
    for m in o:
        a.append(m.get("addr"))
    return a
